### ISSUE TITLE
Excessive waits when listening to events through a websocket

### ISSUE THREAD

**Issue by poolpitako** [ethereum/web3.py#3174]
* Version: 6.12.0
* Python: 3.12
* OS: osx/linux/win
* `pip freeze` output

```
aiohttp==3.9.1
aiosignal==1.3.1
annotated-types==0.6.0
anyio==3.7.1
asyncpg==0.29.0
attrs==23.1.0
beautifulsoup4==4.12.2
bitarray==2.8.5
cachetools==5.3.2
certifi==2023.11.17
charset-normalizer==3.3.2
click==8.1.7
cytoolz==0.12.2
eth-abi==4.2.1
eth-account==0.10.0
eth-hash==0.5.2
eth-keyfile==0.7.0
eth-keys==0.4.0
eth-rlp==1.0.0
eth-typing==3.5.2
eth-utils==2.3.1
fastapi==0.104.1
frozenlist==1.4.0
h11==0.14.0
hexbytes==0.3.1
idna==3.6
jsonschema==4.20.0
jsonschema-specifications==2023.11.2
lru-dict==1.2.0
multidict==6.0.4
parsimonious==0.9.0
pipdeptree==2.13.1
proto-plus==1.23.0
protobuf==4.25.1
pyasn1==0.5.1
pyasn1-modules==0.3.0
pycryptodome==3.19.0
pydantic==2.5.2
pydantic-settings==2.0.3
pydantic_core==2.14.5
python-dotenv==1.0.0
pyunormalize==15.1.0
PyYAML==6.0.1
referencing==0.32.0
regex==2023.10.3
requests==2.31.0
rlp==4.0.0
rpds-py==0.13.2
rsa==4.9
setuptools==69.0.2
six==1.16.0
sniffio==1.3.0
soupsieve==2.5
starlette==0.27.0
toolz==0.12.0
typing_extensions==4.9.0
urllib3==2.1.0
uvicorn==0.23.2
web3==6.12.0
websockets==12.0
yarl==1.9.4
```

### What was wrong?

When user does `listen_to_websocket()`, web3.py ends up doing:
```
response = await self._provider._ws_recv(timeout=0.5)
```

This means that web3.py will wait 0.5 secs for a possible event since websockets' `recv()` will block.
We noticed the issue since we are running a web server with several ws connections to different RPCs. Each ws connection have different subscriptions.

In our tests, we were seeing events arriving several seconds after the websockets library was getting the frame.

### How can it be fixed?

As mentioned in
https://github.com/python-websockets/websockets/issues/162

instead of timeout on websockets' `recv()`, web3.py can check the `websocket.messages` queue directly for event listening.

**Comment by fselmo** [comment]
~~Hey @poolpitako. I like this change a lot. I'm going to take a look at implementing it and adding some good tests around it. Thanks for the submission~~

---
**edit: On further review, this does not seem like a point of friction at the moment. More details are needed, see my comment below.**

**Comment by fselmo** [comment]
@poolpitako I hadn't looked closely here but you'll need to provide a bit more context, including a reproducible code snippet if we are to try and debug what you're experiencing. 

The line you posted from our code does _not_ block the `recv()` call for 0.5 seconds if there are messages in the queue. It only prevents us from listening to an empty queue for longer than 500ms. If messages are present it _will_ return the next message. 

(~~You sent an answer from 2017 that doesn't use the _async_ version of the `websockets` library, the version that is implemented in `WebsocketProviderV2`. Instead it attempts to create an async context using what looks like the then synchronous version of the library. You can see that the `recv()` does not have an await in that example.~~)

[I believe this is irrelevant since they are reaching into the cache directly. It could be the threading or the async version of the library and it actually looks like it's the async version. I stand corrected.]

This is also a 500ms timeout because if no messages come through in 500ms, we will check our internal cache in case any other running processes received the message that our current call wanted and cached it since it may have not been the message that particular call wanted. Does that make sense?

If you're experiencing seconds of wait time, something else is likely going on. I cannot reproduce this on my end as I can subscribe to `newPendingTransactions`, which floods messages in very quickly, with debug logging, and I can see that as a message comes in, it is retrieved and processed. I can even `asyncio.gather` tasks concurrently and the internal caching system works well in most cases, though that is not the recommended use for a websocket provider - but we do provide some flexibility for concurrent tasks in the case that the use case is small enough with not so much latency expected.

Please provide more context along with a reproducible sample code and any debug logs you're seeing that may actually point to this line being your point of friction. Best of luck.

**Comment by fselmo** [comment]
I do see they are reaching into the queue directly but I'm not seeing this as a method anymore in that library. I'm also hesitant that it wouldn't return a message if it's there. I don't think I'm understanding your issue correctly. `await websockets.recv()` will return a message if one is present so it's only blocking in the sense that if nothing is present it will block (which is why the timeout is there, so it doesn't listen indefinitely to an empty queue).

Can you describe a bit more your full context? I'll try to poke around on my end until then to see if I can reproduce a point of friction.

**Comment by fselmo** [comment]
As I mentioned above, subscribing to `newPendingTransactions`, which floods the connection with messages, is successful on my end. I also have a decent amount of success `asyncio.gather()`-ing concurrent tasks while listening to a flood of subscriptions with very little latency. Below I also tested what would happen if we have a response waiting each time `await asyncio.wait_for(ws.recv(), timeout=0.5)` is called (since this is the line the one you pasted here actually ends up calling). Even if the timeout is increased. It shows this `timeout` has nothing to do with any blocks, as long as there is a message present.

For simplicity, here's an example that shows if there is a message in the websocket, it will be returned no matter the timeout.

```py
MOCKED_NEW_HEADS_RESPONSE = {
    "jsonrpc": "2.0",
    "method": "eth_subscription",
    "id": "1",
    "params": {
        "subscription": "THIS_WILL_BE_DYNAMICALLY_SUBSTITUTED",
        "result": {
            "baseFeePerGas": "0x3b9aca00",
            "extraData": "0x",
            "gasLimit": "0x1c9c380",
            "gasUsed": "0x1ec30",
            "size": "0x734",
            "timestamp": "0x2dfdc1c3e",
            "hash": "0x759bc3c1221beb27a7074dbf33faded2276b04df6aaf225d51c426b8c481e935",
            "logsBloom": "0x00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",  # noqa: E501
            "miner": "0x0deadbeefdeadbeef72211ad2C21deadbeeffeed",
            "mixHash": "0xfa167946892d77a2fe116fa81c53c18cfd163889cb08f74c23415c8d13762e1b",
            "nonce": "0x0",
            "number": "0x2dfdc1c3e",
            "parentHash": (
                "0xfe7bde5fac0a5b023504e4abba1165101ca7fe724e6ad29d0c5f122e6bb656fe"
            ),
            "sha3Uncles": (
                "0x1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347"
            ),
            "difficulty": "0x0",
            "receiptsRoot": (
                "0x5ffefe7b059459dd6e1bdecba354eafa33fa0ae619383658bac39b817fdebeef"
            ),
            "stateRoot": (
                "0xfeed54faf6bb151112970b460118ed790197d69df7fcdd79801d120851723c88"
            ),
            "totalDifficulty": "0x9bf9a3",
            "transactionsRoot": (
                "0xafeedbeef5f30efc9baf28e49271302e9b569cfa1e8e9d9512360eb8e7c667a7"
            ),
            "transactions": [
                "0xfcdabb5af17478051493039875221e21282d64b4ed4cf9ba38dff222ecde2f88",
                "0x2636dba09e226a39ad90df7cc43c6a3b353ef6f3ba754f33507a3c29e47467eb",
            ],
            "withdrawalsRoot": (
                "0xfeed23dfefd5340fddc3b23bd6c84a32b8ea964ba16dc4c961ed7caa494cfefe"
            ),
            "withdrawals": [
                {
                    "index": "0x3090f7",
                    "validatorIndex": "0xbfe1",
                    "address": "0xfe102287c050e5ba072211ad2C213eb5dae4feed",
                    "amount": "0x3f695",
                },
                {
                    "index": "0x309103",
                    "validatorIndex": "0xbfed",
                    "address": "0xf22E180C050E5AB072211AD2C213EB5AEE4DF123",
                    "amount": "0x3f695",
                },
            ],
        },
    },
}


async def ws_recv_mock():
    # return MOCKED_NEW_HEADS_RESPONSE as bytes
    return json.dumps(MOCKED_NEW_HEADS_RESPONSE).encode()


async def wsv2():
    async with AsyncWeb3.persistent_websocket(WebsocketProviderV2(WS_URI)) as w3:
        # subscribe to newHeads
        await w3.eth.subscribe("newHeads")

        is_mocking = False
        async for message in w3.ws.listen_to_websocket():
            if not is_mocking:
                # At the first message, get the sub_id for the actual `newHeads`
                # subscription and add it to our MOCKED_NEW_HEADS_RESPONSE. This will
                # allow us to properly process each message since we have the
                # information for the request saved, along with any formatters for this
                # subscription's messages.
                actual_sub_id = message["subscription"]
                MOCKED_NEW_HEADS_RESPONSE["params"]["subscription"] = actual_sub_id

                # replace only the underlying websockets recv() function with
                # our mock that will return a block right away every time
                w3.provider._ws.recv = ws_recv_mock
                is_mocking = True  # begin returning a block right away at next message
            else:
                # this should print the mocked block every time and there won't be a
                # timeout because the block is always returned right away
                print(f"message: {message}\n")

asyncio.run(wsv2())
```

Here, we subscribe to `newHeads`, at the first message we stub the subscription id into our mocked response so we know how to process it correctly, and we start mocking after the first message. The mocked method will return a mocked response right away every time and no matter how much I increase the `timeout` in the line you mentioned, the block is returned each time because there is actually something behind the await, waiting to be returned.

I'd love to keep chatting about this, especially if there is a friction point elsewhere that we should address. However, I don't believe the friction point is where you originally mentioned. If this goes idle for a few days we will likely close this as it doesn't seem to be related to the original submitted issue.

Best of luck.

**Comment by fselmo** [comment]
@poolpitako this did make me think about this logic a bit more though and I do think it was a bit inefficient. I put up a PR to run concurrent tasks to check the cache and the socket and return when the winning task finds it instead. Take a look over at #3176 if you're curious.

I'm going to close this for now but please open another ticket if you are still seeing issues and let's try to pinpoint where the friction is coming from.

**Comment by poolpitako** [comment]
@fselmo sorry for the delay response, I have been traveling.
The root of the issue is what you mentioned here:

> The line you posted from our code does not block the recv() call for 0.5 seconds if there are messages in the queue. It only prevents us from listening to an empty queue for longer than 500ms. If messages are present it will return the next message.

In my code I am doing subscriptions of events which do not happen on every block. An easy repro code would be listening to `add_liquidity` events for different curve pools in different listeners. Let's say that in block 100 there is a single curve pool which emits an `add_liquidity` event and you have 10 different listeners. Messages in the queue would be empty for most of the listeners and the user will be waiting for (10-1)*timeout seconds in the worst case scenario.

**Comment by fselmo** [comment]
@poolpitako I see, thanks for the reply. Is there any way you can test against the PR branch I linked above, #3176, and let us know if the problem persists?

**Comment by poolpitako** [comment]
> @poolpitako I see, thanks for the reply. Is there any way you can test against the PR branch I linked above, #3176, and let us know if the problem persists?

yep, problem persist.
Should I create a smaller version of my codebase?

**Comment by fselmo** [comment]
Please. Otherwise it will be hard to debug.

**Comment by poolpitako** [comment]
@fselmo here's something I hacked together: https://gist.github.com/poolpitako/a38274adc9cd8a68466104daffbbe683

You need to set `WSS_RPC_URL` env var with your wss url.

Example listens to `transfer`, `deposit` and `approval` from WETH contract through different wss connections for 1 minute.

To better understand the problem, add a log in the timeout except inside web3.py code.
```python
                    async with self._provider._ws_lock:
                        try:
                            # keep timeout low but reasonable to check both the cache
                            # and the websocket connection for new responses
                            response = await self._provider._ws_recv(timeout=0.5)
                        except asyncio.TimeoutError:
                            self.logger.debug("TIMEOUT")
```

Here's a gist of the output

```
DEBUG:web3.RequestManager:12:24:47.393:TIMEOUT
DEBUG:web3.RequestManager:12:24:47.894:TIMEOUT
DEBUG:web3.RequestManager:12:24:48.394:TIMEOUT
DEBUG:web3.RequestManager:12:24:48.894:TIMEOUT
DEBUG:web3.RequestManager:12:24:49.395:TIMEOUT
DEBUG:web3.RequestManager:12:24:49.895:TIMEOUT
DEBUG:web3.RequestManager:12:24:50.395:TIMEOUT
DEBUG:listener_<__main__.EventListener object at 0x1132f2c60>:12:24:50.445:Event arrival timestamp: 2023-12-22 15:24:50.445285+00:00 on block 18842258
DEBUG:web3.RequestManager:12:24:50.945:TIMEOUT
DEBUG:listener_<__main__.EventListener object at 0x113480680>:12:24:50.946:Event arrival timestamp: 2023-12-22 15:24:50.946105+00:00 on block 18842258
DEBUG:listener_<__main__.EventListener object at 0x1132f2c60>:12:24:50.946:Event arrival timestamp: 2023-12-22 15:24:50.946483+00:00 on block 18842258
DEBUG:web3.RequestManager:12:24:51.446:TIMEOUT
DEBUG:listener_<__main__.EventListener object at 0x113480680>:12:24:51.447:Event arrival timestamp: 2023-12-22 15:24:51.447350+00:00 on block 18842258
DEBUG:listener_<__main__.EventListener object at 0x1132f2c60>:12:24:51.447:Event arrival timestamp: 2023-12-22 15:24:51.447823+00:00 on block 18842258
DEBUG:web3.RequestManager:12:24:51.948:TIMEOUT
DEBUG:listener_<__main__.EventListener object at 0x113480680>:12:24:51.948:Event arrival timestamp: 2023-12-22 15:24:51.948853+00:00 on block 18842258
DEBUG:listener_<__main__.EventListener object at 0x1132f2c60>:12:24:51.949:Event arrival timestamp: 2023-12-22 15:24:51.949308+00:00 on block 18842258
DEBUG:web3.RequestManager:12:24:52.449:TIMEOUT
DEBUG:listener_<__main__.EventListener object at 0x113480680>:12:24:52.450:Event arrival timestamp: 2023-12-22 15:24:52.450049+00:00 on block 18842258
DEBUG:listener_<__main__.EventListener object at 0x1132f2c60>:12:24:52.450:Event arrival timestamp: 2023-12-22 15:24:52.450421+00:00 on block 18842258
DEBUG:web3.RequestManager:12:24:52.950:TIMEOUT
DEBUG:listener_<__main__.EventListener object at 0x113480680>:12:24:52.951:Event arrival timestamp: 2023-12-22 15:24:52.951305+00:00 on block 18842258
```

See how the timestamp keeps increasing for events in the same block.

You can use this sample code to test against your PR.

**Comment by fselmo** [comment]
@poolpitako Thank you for the gist! It will help a ton.

**Comment by fselmo** [comment]
@poolpitako. I definitely see the issue. It can be mitigated for now by using the same connection to the websocket for all three listeners, but this is _not_ ideal. Each connection is yielding to the next within the event loop since they share the loop, back and forth, and this triggers the `0.5` second timeout quite often.

I have a few different implementations working locally but I'd like to spend a bit more time figuring out an ideal way, add tests, etc. What's more important for now is that you aren't blocked on your implementation. Are you able to get around this issue for now?

**Comment by poolpitako** [comment]
> Are you able to get around this issue for now?

Thanks! We are not blocked and we are happy to wait for a fix.
Do you believe you will need to break the API to fix it?

**Comment by fselmo** [comment]
>  Do you believe you will need to break the API to fix it?

I don't believe that will be necessary.
